src/machine/machine_rp2040_pwm.go

```diff
@@ -256,21 +255,34 @@ func (pwm *pwmGroup) setDivMode(mode uint32) {
 // setPeriod sets the pwm peripheral period (frequency). Calculates DIV_INT and sets it from following equation:
 //  cycles = (TOP+1) * (CSRPHCorrect + 1) * (DIV_INT + DIV_FRAC/16)
 // where cycles is amount of clock cycles per PWM period.
-func (pwm *pwmGroup) setPeriod(period uint64) {
-	targetPeriod := uint32(period)
-	periodPerCycle := getPeriod()
+func (pwm *pwmGroup) setPeriod(period uint32) {
+	periodPerCycle := cpuPeriod()
 	top := pwm.getWrap()
 	phc := pwm.getPhaseCorrect()
-	_, frac := pwm.getClockDiv()
+	// We first modify TOP if desired period is outside divider threshold bounds.
+	limitPeriod := (1 + phc) * periodPerCycle * (1 + top)
+	switch {
+	case period < limitPeriod:
+		// Desired period is below reachable threshold.
+		top = period/((1+phc)*periodPerCycle) - 1
+		pwm.setWrap(uint16(top))
+	case period > 255*limitPeriod:
+		// Desired period is above reachable threshold.
+		top = (period/(256*(1+phc)*periodPerCycle) - 1)
+		if top > 0xffff {
+			top = 0xffff // TOP max representable value is uint16
+		}
+		pwm.setWrap(uint16(top))
+	}
+
 	// clearing above expression:
-	//  DIV_INT = cycles / ( (TOP+1) * (CSRPHCorrect+1) ) - DIV_FRAC/16
+	//  DIV_INT + DIV_FRAC/16 = cycles / ( (TOP+1) * (CSRPHCorrect+1) )  // DIV_FRAC/16 is always 0 in this equation
 	// where cycles must be converted to time:
 	//  target_period = cycles * period_per_cycle ==> cycles = target_period/period_per_cycle
-	Int := targetPeriod/((1+phc)*periodPerCycle*(1+top)) - frac/16
-	if Int > 0xff {
-		Int = 0xff
-	}
-	pwm.setClockDiv(uint8(Int), 0)
+	rhs := 16 * uint64(period) / uint64((1+phc)*periodPerCycle*(1+top)) // right-hand-side of equation, scaled so frac is not divided
+	Int := (rhs / 16) & 0xff
```

Comment:
Kinda of a nit-pick, but I would have named this `whole` or something instead of `Int` which is probably only exported due to name collision with `int` which seems like a call for this to be named something else. :wink:

Reply:
Your wish is my command :wink: , also good call